src/hotspot/os/linux/cgroupUtil_linux.cpp

```diff
@@ -46,3 +47,99 @@ int CgroupUtil::processor_count(CgroupCpuController* cpu_ctrl, int host_cpus) {
   log_trace(os, container)("OSContainer::active_processor_count: %d", result);
   return result;
 }
+
+CgroupMemoryController* CgroupUtil::adjust_controller(CgroupMemoryController* mem) {
+  if (!mem->needs_hierarchy_adjustment()) {
+    // nothing to do
+    return mem;
+  }
+  log_trace(os, container)("Adjusting controller path for memory: %s", mem->subsystem_path());
+  assert(mem->cgroup_path() != nullptr, "invariant");
+  char* orig = os::strdup(mem->cgroup_path());
+  char* cg_path = os::strdup(orig);
+  char* last_slash;
+  julong phys_mem = os::Linux::physical_memory();
+  jlong limit = mem->read_memory_limit_in_bytes(phys_mem);
+  bool path_iterated = false;
+  while (limit < 0 && (last_slash = strrchr(cg_path, '/')) != cg_path) {
```

Comment:
Is it guaranteed that cg_path starts with / ? otherwise, you have a write-to-NULL below. maybe assert that `cg_path[0] == '/'`

Reply:
Yes, that should be guaranteed. The assertion is a good idea. Added.